### PR TITLE
Merge old statistic records into one

### PR DESCRIPTION
**Pull request by belmirp**
This job will be executed once per day at the time defined in the appsettings.

It will look for daily non-flagged statistic records that are older than 7 days, and merge them into one statistic record with `Window = DailyMerged`. If there is only one record per account per day, that record is only modified to have `Window = DailyMerged`.

Here are some examples of the process:

MergingOldStatistics_Example1
MergingOldStatistics_Example2
MergingOldStatistics_Example3

### CODE REVIEW

File: UT4MasterServer/Models/StatisticBase.cs [hunk 1]
```diff
@@ -516,109 +516,130 @@ public List<string> Validate()
...
 		{
 			flaggedFields.Add(nameof(MinigunKills));
+			flaggedFields.Add(nameof(Kills));
 		}
 		// Minigun Shard kills are reported if they are more than the actual kills
-		if (MinigunShardKills.HasValue && Kills.HasValue && MinigunShardKills.Value > Kills.Value)
+		if (MinigunShardKills.HasValue && (!Kills.HasValue || MinigunShardKills.Value > Kills.Value))
 		{
 			flaggedFields.Add(nameof(MinigunShardKills));
+			flaggedFields.Add(nameof(Kills));
```
Comment (Saibamen) [inline UT4MasterServer/Models/StatisticBase.cs:582]: Maybe add a helper method, for example `AddKillsFor` to reduce code duplication

Comment (belmirp) [reply UT4MasterServer/Models/StatisticBase.cs:582]: This part is adding only name of the property to the list of flagged properties. I am not sure for what exactly would you add extension method?

Comment (Saibamen) [reply UT4MasterServer/Models/StatisticBase.cs:582]: To reduce code. I updated my comment with code example

Comment (belmirp) [reply UT4MasterServer/Models/StatisticBase.cs:582]: You are not going to reduce anything. You will replace this:

image

with this:
image

Comment (belmirp) [reply UT4MasterServer/Models/StatisticBase.cs:582]: If the line count is important, I could do it like this:
image

File: UT4MasterServer/Services/StatisticsService.cs [hunk 2]
```diff
@@ -249,6 +249,125 @@ public async Task CreateAccountStatisticsAsync(EpicID accountID, OwnerType _, St
...
+
+		foreach (var statisticsGroup in statisticsGrouped)
+		{
+			// Statistics that have multiple records per day should be merged together or merged with already merged statistics
+			if (statisticsGroup.Count > 1)
+			{
+				var statisticsToMerge = new List<Statistic>(statisticsGroup.Statistics);
+
+				if (alreadyMergedStatistics.FirstOrDefault(f => f.AccountID == statisticsGroup.AccountId &&
+																f.CreatedAt.Date == statisticsGroup.CreatedAt.Date) is { } alreadyMergedStatistic)
```
Comment (Saibamen) [inline UT4MasterServer/Services/StatisticsService.cs:302]: Fix indentation

File: UT4MasterServer/Models/StatisticBase.cs (hunk 1 above)
Comment (Saibamen) [reply UT4MasterServer/Models/StatisticBase.cs:582]: But IMHO using `AddRange` will be better for code optimization.

image

(Remember to add comma at the end for better git diff view and PR code review experience in the future

Comment (belmirp) [reply UT4MasterServer/Models/StatisticBase.cs:582]: The problem with AddRange is that you have to instantiate unnecessary array/list each time. Using Add in multiple lines doesn't allocate anything additional.

File: UT4MasterServer/Services/ApplicationBackgroundService.cs [hunk 3]
```diff
@@ -1,66 +1,87 @@
...
 
-		public void Dispose()
+	/// <summary>
+	/// This job is executed each day, and it merges statistics older than 7 days into single record per day per account
+	/// </summary>
+	private async Task MergeOldStatisticsAsync(IServiceScope scope)
+	{
+		var currentTime = TimeZoneInfo.ConvertTime(DateTimeOffset.UtcNow, TimeZoneInfo.FindSystemTimeZoneById(settings.MergeOldStatisticsTimeZone));
+		if (currentTime.Hour == settings.MergeOldStatisticsHour &&
+		   (!lastDateMergeOldStatisticsExecuted.HasValue || lastDateMergeOldStatisticsExecuted.Value.Date != currentTime.Date))
```
[the remark this answers is not in the document]
Comment (belmirp) [reply UT4MasterServer/Services/ApplicationBackgroundService.cs:77]: I actually added that to be in parallel with the condition above. What do you suggest, to stick with the tabs, and have something like this:
image

or could be worse like this:
image

File: UT4MasterServer/Models/StatisticBase.cs (hunk 1 above)
Comment (timiimit) [reply UT4MasterServer/Models/StatisticBase.cs:582]: How about just having something like this (provided just as pseudocode):
```cs
	public void ValidateProperty(string name, int? value)
	{
		if (value.HasValue && Kills.HasValue && value.Value > Kills.Value)
		{
			flaggedFields.Add(name);
			flaggedFields.Add(nameof(Kills));
		}
	}
```

Comment (timiimit) [reply UT4MasterServer/Models/StatisticBase.cs:582]: This would reduce number of lines significantly imo

Comment (timiimit) [reply UT4MasterServer/Models/StatisticBase.cs:582]: Then create an array with all name/value pairs and just loop though all to reduce perhaps even more

File: UT4MasterServer/Services/ApplicationBackgroundService.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (timiimit) [reply UT4MasterServer/Services/ApplicationBackgroundService.cs:77]: You can create additional variable separating conditions apart

File: UT4MasterServer/Models/StatisticBase.cs (hunk 1 above)
Comment (belmirp) [reply UT4MasterServer/Models/StatisticBase.cs:582]: Not all rules are the same. I wouldn't make multiple methods out of one just for the sake of lines.

Comment (timiimit) [reply UT4MasterServer/Models/StatisticBase.cs:582]: All weapon stats seem the same. Having ~5 extra methods with different conditions is definitely something i'd prefer over this long unreadable mess. Its not for the sake of lines, its for the sake of maintainability and ease of reading.

File: UT4MasterServer/Services/StatisticsService.cs [hunk 4]
```diff
@@ -249,6 +249,125 @@ public async Task CreateAccountStatisticsAsync(EpicID accountID, OwnerType _, St
...
```
Comment (timiimit) [inline UT4MasterServer/Services/StatisticsService.cs:302]: Indentation

File: UT4MasterServer/Services/StatisticsService.cs [hunk 5]
```diff
@@ -249,6 +249,125 @@ public async Task CreateAccountStatisticsAsync(EpicID accountID, OwnerType _, St
...
+
+				statisticIdsToDelete.AddRange(mergedStatisticsIds);
+
+				logger.LogInformation("Merged the following statistics into one record: {StatisticIds}.", string.Join(", ", mergedStatisticsIds));
+			}
+			// Statistics that have one record per day should be either modified or merged with already merged statistics
+			else
+			{
+				if (alreadyMergedStatistics.FirstOrDefault(f => f.AccountID == statisticsGroup.AccountId &&
+																f.CreatedAt.Date == statisticsGroup.CreatedAt.Date) is { } alreadyMergedStatistic)
```
Comment (timiimit) [inline UT4MasterServer/Services/StatisticsService.cs:322]: Here too

File: UT4MasterServer/Models/StatisticBase.cs (hunk 1 above)
Comment (belmirp) [reply UT4MasterServer/Models/StatisticBase.cs:582]: If that is what you prefer, I will extract similar logic to separate methods.

Comment (Saibamen) [reply UT4MasterServer/Models/StatisticBase.cs:582]: I run a simple BenchmarkDotNet benchmark and `DoubleAdd` is better

```
|    Method |       Mean |     Error |     StdDev |     Median |        Min |         Max | Ratio | RatioSD | Allocated | Alloc Ratio |
|---------- |-----------:|----------:|-----------:|-----------:|-----------:|------------:|------:|--------:|----------:|------------:|
| DoubleAdd |   510.7 ns |  68.49 ns |   929.4 ns |   400.0 ns |   300.0 ns | 24,700.0 ns |  1.00 |    0.00 |     640 B |        1.00 |
|  AddRange | 2,627.3 ns | 187.31 ns | 2,541.9 ns | 2,100.0 ns | 1,500.0 ns | 25,600.0 ns |  6.21 |    5.98 |     680 B |        1.06 |
```

Benchmark code:

```C#
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Engines;
using BenchmarkDotNet.Running;

namespace Benchmark;

[SimpleJob(RunStrategy.Monitoring, launchCount: 2, warmupCount: 5, iterationCount: 1000)]
[MemoryDiagnoser]
[MinColumn, MaxColumn, MedianColumn]
public class AddTwiceVsAddRange
{
	public int? Wins { get; set; }
	public int? Losses { get; set; }

	private List<string> flaggedFields;

	[GlobalSetup]
	public void GlobalSetup()
	{
		flaggedFields = new List<string>();
	}

	[Benchmark(Baseline = true)]
	public List<string> DoubleAdd()
	{
		flaggedFields.Add(nameof(Wins));
		flaggedFields.Add(nameof(Losses));
		return flaggedFields;
	}

	[Benchmark]
	public List<string> AddRange()
	{
		flaggedFields.AddRange(new[]
		{
			nameof(Wins),
			nameof(Losses),
		});

		return flaggedFields;
	}
}

public class Program
{
	public static void Main(string[] args)
	{
		var summary = BenchmarkRunner.Run<AddTwiceVsAddRange>();
	}
}
```

Comment (belmirp) [reply UT4MasterServer/Models/StatisticBase.cs]: Extracted logic to multiple methods.

File: UT4MasterServer/Services/StatisticsService.cs (hunk 4 above)
Comment (belmirp) [reply UT4MasterServer/Services/StatisticsService.cs:302]: Indentation is OK, just not displayed by GitHub correctly. You can check it by viewing the whole file:
image

File: UT4MasterServer/Services/StatisticsService.cs (hunk 5 above)
Comment (belmirp) [reply UT4MasterServer/Services/StatisticsService.cs:322]: Same as above.

File: UT4MasterServer/Models/StatisticBase.cs [hunk 6]
```diff
@@ -1110,4 +779,176 @@ public List<string> Validate()
...
+			flaggedFields.Add(statisticName);
+			flaggedFields.Add(nameof(TimePlayed));
+		}
+	}
+
+	private void ValidateMultiAndSpreeKillStatistic(int? value, int multiplier, string statisticName, List<string> flaggedFields)
+	{
+		if (value.HasValue)
+		{
+			if (!Kills.HasValue || (Kills.HasValue && value.Value * multiplier > Kills.Value))
```
Comment (Saibamen) [inline UT4MasterServer/Models/StatisticBase.cs:924]: Merge it into one IF statement or return like this to reduce nesting:

```C#
if (!value.HasValue)
{
	return;
}
```

File: UT4MasterServer/Models/StatisticBase.cs [hunk 7]
```diff
@@ -1110,4 +779,176 @@ public List<string> Validate()
...
+				flaggedFields.Add(statisticName);
+				flaggedFields.Add(nameof(Kills));
+			}
+		}
+	}
+
+	private void ValidateTimeStatistic(double? value, string statisticName, List<string> flaggedFields)
+	{
+		if (value.HasValue)
+		{
```
Comment (Saibamen) [inline UT4MasterServer/Models/StatisticBase.cs:935]: Same here

**Review by Saibamen — CHANGES_REQUESTED** [review]

File: UT4MasterServer/Models/StatisticBase.cs [hunk 8]
```diff
@@ -1089,4 +779,176 @@ public List<string> Validate()
...
+		ValidateIfBothStatisticsExist(MatchesPlayed, MatchesQuit, nameof(MatchesPlayed), nameof(MatchesQuit), flaggedFields);
+		ValidateIfBothStatisticsExist(Wins, Losses, nameof(Wins), nameof(Losses), flaggedFields);
+
+		// Maximum time for custom game is 60 minutes (1 minute is added just to be sure)
+		if (TimePlayed.HasValue && TimePlayed.Value > (60 * 60) + 60)
+		{
+			flaggedFields.Add(nameof(TimePlayed));
+		}
+		if (Kills.HasValue)
+		{
```
Comment (Saibamen) [inline UT4MasterServer/Models/StatisticBase.cs:830]:
```suggestion
		}

		if (Kills.HasValue)
		{
```

File: UT4MasterServer/Models/StatisticBase.cs (hunk 6 above)
Comment (belmirp) [reply UT4MasterServer/Models/StatisticBase.cs]: Done.

File: UT4MasterServer/Models/StatisticBase.cs (hunk 7 above)
Comment (belmirp) [reply UT4MasterServer/Models/StatisticBase.cs]: Done.

**Review by timiimit — APPROVED** [review]

**Comment by belmirp** [comment]
Related: [#22](https://github.com/timiimit/UT4MasterServer/issues/22)

**Pull request merged**